File: applications/rtkregularizedconjugategradient/rtkregularizedconjugategradient.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkregularizedconjugategradient_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkRegularizedConjugateGradientConeBeamReconstructionFilter.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"

#ifdef RTK_USE_CUDA
#  include "itkCudaImage.h"
#  include "rtkCudaConstantVolumeSource.h"
#endif
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkregularizedconjugategradient, args_info);

  using OutputPixelType = float;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, 3>;
#else
  using OutputImageType = itk::Image<OutputPixelType, 3>;
#endif

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkregularizedconjugategradient>(reader, args_info);

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Create input: either an existing volume read from a file or a blank image
  itk::ImageSource<OutputImageType>::Pointer inputFilter;
  if (args_info.input_given)
  {
    // Read an existing image to initialize the volume
    auto inputReader = itk::ImageFileReader<OutputImageType>::New();
    inputReader->SetFileName(args_info.input_arg);
    inputFilter = inputReader;
  }
  else
  {
    // Create new empty volume
    using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
    auto constantImageSource = ConstantImageSourceType::New();
    rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkregularizedconjugategradient>(
      constantImageSource, args_info);
    inputFilter = constantImageSource;
  }
  TRY_AND_EXIT_ON_ITK_EXCEPTION(inputFilter->Update())
  inputFilter->ReleaseDataFlagOn();

  // Read weights if given, otherwise default to weights all equal to one
  itk::ImageSource<OutputImageType>::Pointer weightsSource;
  if (args_info.weights_given)
  {
    auto weightsReader = itk::ImageFileReader<OutputImageType>::New();
    weightsReader->SetFileName(args_info.weights_arg);
    weightsSource = weightsReader;
  }
  else
  {
    auto constantWeightsSource = rtk::ConstantImageSource<OutputImageType>::New();

    // Set the weights to be like the projections
    TRY_AND_EXIT_ON_ITK_EXCEPTION(reader->UpdateOutputInformation())
    constantWeightsSource->SetInformationFromImage(reader->GetOutput());
    constantWeightsSource->SetConstant(1.0);
    weightsSource = constantWeightsSource;
  }

  // Read Support Mask if given
  OutputImageType::Pointer supportmaskSource;
  if (args_info.mask_given)
  {
    TRY_AND_EXIT_ON_ITK_EXCEPTION(supportmaskSource = itk::ReadImage<OutputImageType>(args_info.mask_arg))
  }

  // Set the forward and back projection filters to be used
  using ConjugateGradientFilterType = rtk::RegularizedConjugateGradientConeBeamReconstructionFilter<OutputImageType>;
  auto regularizedConjugateGradient = ConjugateGradientFilterType::New();
  SetForwardProjectionFromGgo(args_info, regularizedConjugateGradient.GetPointer());
  SetBackProjectionFromGgo(args_info, regularizedConjugateGradient.GetPointer());
  regularizedConjugateGradient->SetInputVolume(inputFilter->GetOutput());
  regularizedConjugateGradient->SetInputProjectionStack(reader->GetOutput());
  regularizedConjugateGradient->SetInputWeights(weightsSource->GetOutput());
  regularizedConjugateGradient->SetGeometry(geometry);
  regularizedConjugateGradient->SetMainLoop_iterations(args_info.niter_arg);
  regularizedConjugateGradient->SetCudaConjugateGradient(!args_info.nocudacg_flag);
  regularizedConjugateGradient->SetDisableDisplacedDetectorFilter(args_info.nodisplaced_flag);
  if (args_info.mask_given)
  {
    regularizedConjugateGradient->SetSupportMask(supportmaskSource);
  }

  // Positivity
  if (args_info.nopositivity_flag)
    regularizedConjugateGradient->SetPerformPositivity(false);
  else
    regularizedConjugateGradient->SetPerformPositivity(true);

  if (args_info.gammalaplacian_given)
    regularizedConjugateGradient->SetGamma(args_info.gammalaplacian_arg);
  if (args_info.tikhonov_given)
    regularizedConjugateGradient->SetTikhonov(args_info.tikhonov_arg);

  // Spatial TV
  if (args_info.gammatv_given)
  {
    regularizedConjugateGradient->SetGammaTV(args_info.gammatv_arg);
    regularizedConjugateGradient->SetTV_iterations(args_info.tviter_arg);
    regularizedConjugateGradient->SetPerformTVSpatialDenoising(true);
  }
  else
    regularizedConjugateGradient->SetPerformTVSpatialDenoising(false);

  // Spatial wavelets
  if (args_info.threshold_given)
  {
    regularizedConjugateGradient->SetSoftThresholdWavelets(args_info.threshold_arg);
    regularizedConjugateGradient->SetOrder(args_info.order_arg);
    regularizedConjugateGradient->SetNumberOfLevels(args_info.levels_arg);
    regularizedConjugateGradient->SetPerformWaveletsSpatialDenoising(true);
  }
  else
    regularizedConjugateGradient->SetPerformWaveletsSpatialDenoising(false);

  // Sparsity in image domain
  if (args_info.soft_given)
  {
    regularizedConjugateGradient->SetSoftThresholdOnImage(args_info.soft_arg);
    regularizedConjugateGradient->SetPerformSoftThresholdOnImage(true);
  }
  else
    regularizedConjugateGradient->SetSoftThresholdOnImage(false);

  REPORT_ITERATIONS(regularizedConjugateGradient, ConjugateGradientFilterType, OutputImageType)

  TRY_AND_EXIT_ON_ITK_EXCEPTION(regularizedConjugateGradient->Update())

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(regularizedConjugateGradient->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkregularizedconjugategradient/rtkregularizedconjugategradient.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Alternates between conjugate gradient reconstruction and regularization"
    )

    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--niter",
        "-n",
        help="Number of iterations",
        type=int,
        default=5,
    )

    parser.add_argument("--input", "-i", help="Input volume", type=str)
    parser.add_argument(
        "--weights",
        "-w",
        help="Weights file for Weighted Least Squares (WLS)",
        type=str,
    )

    parser.add_argument(
        "--gammalaplacian",
        help="Laplacian regularization weight",
        type=float,
        default=0.0,
    )
    parser.add_argument(
        "--tikhonov",
        help="Tikhonov regularization weight",
        type=float,
        default=0.0,
    )
    parser.add_argument(
        "--cgiter",
        help="Number of conjugate gradient nested iterations",
        type=int,
        default=4,
    )
    parser.add_argument(
        "--nocudacg",
        help="Do not perform conjugate gradient calculations on GPU",
        action="store_true",
    )
    parser.add_argument(
        "--mask",
        "-m",
        help="Apply a support binary mask: reconstruction kept null outside the mask)",
        type=str,
    )
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )

    reg = parser.add_argument_group("Regularization")
    reg.add_argument(
        "--nopositivity",
        help="Do not enforce positivity",
        action="store_true",
    )
    reg.add_argument(
        "--tviter",
        help="Total variation regularization: number of iterations",
        type=int,
        default=10,
    )
    reg.add_argument(
        "--gammatv",
        help="Total variation spatial regularization parameter. The larger, the smoother",
        type=float,
    )
    reg.add_argument(
        "--threshold",
        help="Daubechies wavelets spatial regularization: soft threshold",
        type=float,
    )
    reg.add_argument(
        "--order",
        help="Daubechies wavelets spatial regularization: order of the wavelets",
        type=int,
        default=5,
    )
    reg.add_argument(
        "--levels",
        help="Daubechies wavelets spatial regularization: number of decomposition levels",
        type=int,
        default=3,
    )
    reg.add_argument(
        "--soft",
        help="Soft threshold for image domain sparsity enforcement",
        type=float,
    )

    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtk3Doutputimage_group(parser)
    rtk.add_rtkprojectors_group(parser)
    rtk.add_rtkiterations_group(parser)

    return parser


def process(args_info: argparse.Namespace) -> None:
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    reader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Input volume: file or constant image
    if args_info.input:
        inputFilter = itk.ImageFileReader[OutputImageType].New()
        inputFilter.SetFileName(args_info.input)
    else:
        inputFilter = rtk.ConstantImageSource[OutputImageType].New()
        rtk.SetConstantImageSourceFromArgParse(inputFilter, args_info)

    inputFilter.Update()

    # Weights: file or constant 1 image matching projections
    if args_info.weights:
        weightsSource = itk.ImageFileReader[OutputImageType].New()
        weightsSource.SetFileName(args_info.weights)
    else:
        weightsSource = rtk.ConstantImageSource[OutputImageType].New()
        reader.UpdateOutputInformation()
        weightsSource.SetInformationFromImage(reader.GetOutput())
        weightsSource.SetConstant(1.0)

    # Support mask
    if args_info.mask:
        supportmask = itk.imread(args_info.mask)

    # Set the forward and back projection filters to be used
    if hasattr(itk, "CudaImage") and not args_info.nocudacg:
        CudaImageType = itk.CudaImage[OutputPixelType, Dimension]
        regularizedConjugateGradient = (
            rtk.RegularizedConjugateGradientConeBeamReconstructionFilter[
                CudaImageType
            ].New()
        )
        regularizedConjugateGradient.SetInputVolume(
            itk.cuda_image_from_image(inputFilter.GetOutput())
        )
        regularizedConjugateGradient.SetInputProjectionStack(
            itk.cuda_image_from_image(reader.GetOutput())
        )
        regularizedConjugateGradient.SetInputWeights(
            itk.cuda_image_from_image(weightsSource.GetOutput())
        )
        if args_info.mask:
            regularizedConjugateGradient.SetSupportMask(
                itk.cuda_image_from_image(supportmask)
            )

    else:
        regularizedConjugateGradient = (
            rtk.RegularizedConjugateGradientConeBeamReconstructionFilter[
                OutputImageType
            ].New()
        )

        # Inputs
        regularizedConjugateGradient.SetInputVolume(inputFilter.GetOutput())
        regularizedConjugateGradient.SetInputProjectionStack(reader.GetOutput())
        regularizedConjugateGradient.SetInputWeights(weightsSource.GetOutput())
        if args_info.mask:
            regularizedConjugateGradient.SetSupportMask(supportmask)

    rtk.SetForwardProjectionFromArgParse(args_info, regularizedConjugateGradient)
    rtk.SetBackProjectionFromArgParse(args_info, regularizedConjugateGradient)

    regularizedConjugateGradient.SetGeometry(geometry)
    regularizedConjugateGradient.SetMainLoop_iterations(args_info.niter)
    regularizedConjugateGradient.SetCG_iterations(args_info.cgiter)
    regularizedConjugateGradient.SetCudaConjugateGradient(not args_info.nocudacg)
    regularizedConjugateGradient.SetDisableDisplacedDetectorFilter(
        args_info.nodisplaced
    )
    regularizedConjugateGradient.SetPerformPositivity(not args_info.nopositivity)
    regularizedConjugateGradient.SetGamma(args_info.gammalaplacian)
    regularizedConjugateGradient.SetTikhonov(args_info.tikhonov)

    # TV
    if args_info.gammatv is not None:
        regularizedConjugateGradient.SetGammaTV(args_info.gammatv)
        regularizedConjugateGradient.SetTV_iterations(args_info.tviter)
        regularizedConjugateGradient.SetPerformTVSpatialDenoising(True)
    else:
        regularizedConjugateGradient.SetPerformTVSpatialDenoising(False)

    # Wavelets
    if args_info.threshold is not None:
        regularizedConjugateGradient.SetSoftThresholdWavelets(args_info.threshold)
        regularizedConjugateGradient.SetOrder(args_info.order)
        regularizedConjugateGradient.SetNumberOfLevels(args_info.levels)
        regularizedConjugateGradient.SetPerformWaveletsSpatialDenoising(True)
    else:
        regularizedConjugateGradient.SetPerformWaveletsSpatialDenoising(False)

    # Image-domain soft threshold
    if args_info.soft is not None:
        regularizedConjugateGradient.SetSoftThresholdOnImage(args_info.soft)
        regularizedConjugateGradient.SetPerformSoftThresholdOnImage(True)
    else:
        regularizedConjugateGradient.SetPerformSoftThresholdOnImage(False)

    # Iteration reporting
    rtk.SetIterationsReportFromArgParse(args_info, regularizedConjugateGradient)

    if args_info.verbose:
        print("Reconstructing...")

    regularizedConjugateGradient.Update()

    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")

    # Write
    writer = itk.ImageFileWriter[OutputImageType].New()
    writer.SetFileName(args_info.output)
    writer.SetInput(regularizedConjugateGradient.GetOutput())
    writer.Update()


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
